src/lib/operators/table_scan/column_vs_value_table_scan_impl.cpp

```diff
@@ -127,6 +133,51 @@ void ColumnVsValueTableScanImpl::_scan_dictionary_segment(const BaseDictionarySe
   });
 }
 
+void ColumnVsValueTableScanImpl::_scan_sorted_segment(const BaseSegment& segment, const ChunkID chunk_id,
+                                                      PosList& matches,
+                                                      const std::shared_ptr<const PosList>& position_filter,
+                                                      const OrderByMode ordered_by) const {
+  resolve_data_and_segment_type(segment, [&](const auto type, const auto& typed_segment) {
+    using ColumnDataType = typename decltype(type)::type;
+
+    if constexpr (std::is_same_v<decltype(typed_segment), const ReferenceSegment&>) {
+      Fail("Expected ReferenceSegments to be handled before calling this method");
+    } else {
+      auto segment_iterable = create_iterable_from_segment(typed_segment);
+      segment_iterable.with_iterators(position_filter, [&](auto segment_begin, auto segment_end) {
+        auto sorted_segment_search = detail::SortedSegmentSearch(
+            segment_begin, segment_end, ordered_by, _predicate_condition, type_cast_variant<ColumnDataType>(_value));
+
+        sorted_segment_search.scan_sorted_segment([&](auto begin, auto end) {
+          size_t output_idx = matches.size();
+
+          matches.resize(matches.size() + std::distance(begin, end));
+
+          /**
+           * If the range of matches consists of continuous ChunkOffsets we can speed up the writing
+           * by calculating the offsets based on the first offset instead of calling chunk_offset()
+           * for every match.
+           * ChunkOffsets in position_filter are not necessarily continuous. The same is true for
+           * NotEquals because the result might consist of 2 ranges.
+           */
+          if (position_filter || _predicate_condition == PredicateCondition::NotEquals) {
+            for (; begin != end; ++begin) {
+              matches[output_idx++] = RowID(chunk_id, begin->chunk_offset());
+            }
+          } else {
+            const auto first_offset = begin->chunk_offset();
+            const auto distance = std::distance(begin, end);
+
+            for (auto chunk_offset = 0; chunk_offset < distance; ++chunk_offset) {
+              matches[output_idx++] = RowID(chunk_id, first_offset + chunk_offset);
```

Comment:
Do you have numbers for this?

Reply:
I benchmarked the code again to compare our offset calculation approach to the "standard" approach. It seems that our approach is indeed slower on int segments which are Dictionary encoded.
If there is no enconding, our method is better and gets even better on strings especially if those strings are long. (Had to use logarithmic scaling.)
![image](https://user-images.githubusercontent.com/16731894/54074846-f33f0800-4297-11e9-9f28-b8300abe9b19.png)
![image](https://user-images.githubusercontent.com/16731894/54074850-0651d800-4298-11e9-9fcc-735aa03c19fd.png)
![image](https://user-images.githubusercontent.com/16731894/54074866-41540b80-4298-11e9-9394-da815cebfd02.png)
![image](https://user-images.githubusercontent.com/16731894/54074874-503abe00-4298-11e9-8de2-dd2851a7d252.png)

Reply:
Oh, wow. I wouldn't have expected that much of a difference. We can live with the small overhead for that one case if we get those huge benefits in the others.